libpromises/evalfunction.c

```diff
@@ -6659,6 +6658,72 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
     return (FnCallResult) { FNCALL_SUCCESS, (Rval) { json, RVAL_TYPE_CONTAINER } };
 }
 
+static FnCallResult FnCallClassExprFilter(EvalContext *ctx,
+                                          ARG_UNUSED Policy const *policy,
+                                          ARG_UNUSED FnCall const *fp,
+                                          Rlist const *args)
+{
+    char const *path = RlistScalarValue(args);
+    size_t const col = IntFromString(RlistScalarValue(args->next));
+    char const *delim = RlistScalarValue(args->next->next);
+    bool has_heading = BooleanFromString(RlistScalarValue(args->next->next->next));
+    int const sort_col = IntFromString(RlistScalarValue(args->next->next->next->next));
+
+    Seq *heading_seq = NULL;
+    char *line;
+    JsonElement *json = JsonArrayCreate(50);
+
+    FILE *csv_file = safe_fopen(path, "r");
+    if (csv_file == NULL)
+    {
+        return FnFailure();
+    }
+
+    while ((line = GetCsvLineNext(csv_file)) != NULL)
+    {
+        Seq *seq = SeqParseCsvString(line);
+        if (seq == NULL)
+        {
+            return FnFailure();
+        }
+
+        free(line);
+
+        if (!has_heading && !IsDefinedClass(ctx, SeqAt(seq, col)))
+        {
+            SeqDestroy(seq);
+            continue;
+        }
+
+        SeqRemove(seq, col);
+
+        if (has_heading)
+        {
+            heading_seq = seq;
+            has_heading = false;
+        }
+        else
+        {
+            size_t len = SeqLength(seq);
+
+            JsonElement *json_object = JsonObjectCreate(len);
+            for (size_t i = 0; i < len; i++)
+            {
+                JsonObjectAppendString(json_object,
+                                       SeqAt(heading_seq, i),
+                                       SeqAt(seq, i));
```

Comment:
It seems you would need to range check the heading_seq length against the sequence length for each line. SeqAt() only has asserts so it's `return seq->data[i]` could get out of range.